Algorithm.CSharp/RegressionTests/Universes/ETFConstituentUniverseFilterFunctionRegressionAlgorithm.cs

```diff
@@ -0,0 +1,205 @@
+        /// </summary>
+        public override void Initialize()
+        {
+            SetStartDate(2020, 12, 1);
+            SetEndDate(2021, 1, 31);
+            SetCash(100000);
+
+            UniverseSettings.Resolution = Resolution.Hour;
+            
+            _spy = AddEquity("SPY", Resolution.Hour).Symbol;
+            _aapl = QuantConnect.Symbol.Create("AAPL", SecurityType.Equity, Market.USA);
+            
+            AddUniverse(new ETFConstituentsUniverse(_spy, UniverseSettings, FilterETFs));
+        }
+
+        /// <summary>
+        /// Filters ETFs, performing some sanity checks
+        /// </summary>
+        /// <param name="constituents">Constituents of the ETF universe added above</param>
+        /// <returns>Constituent Symbols to add to algorithm</returns>
+        /// <exception cref="ArgumentException">Constituents collection was not structured as expected</exception>
+        private IEnumerable<Symbol> FilterETFs(IEnumerable<ETFConstituentData> constituents)
+        {
+            var constituentsData = constituents.ToHashSet();
+            var constituentsSymbols = constituentsData.Select(x => x.Symbol).ToList();
+            if (constituentsData.Count == 0)
+            {
+                throw new ArgumentException($"Constituents collection is empty on {UtcTime:yyyy-MM-dd HH:mm:ss.fff}");
+            }
+            if (!constituentsSymbols.Contains(_aapl))
+            {
+                throw new ArgumentException("AAPL is not in the constituents data provided to the algorithm");
+            }
+
+            var aaplData = constituentsData.Single(x => x.Symbol == _aapl);
+            if (aaplData.Weight == 0m)
+            {
+                throw new ArgumentException("AAPL weight is expected to be a non-zero value");
+            }
+
+            _filtered = true;
+            return constituentsSymbols;
+        }
+
+        /// <summary>
+        /// OnData event is the primary entry point for your algorithm. Each new data point will be pumped in here.
+        /// </summary>
+        /// <param name="data">Slice object keyed by symbol containing the stock data</param>
+        public override void OnData(Slice data)
+        {
+            if (!_filtered && data.Bars.Count != 0 && data.Bars.ContainsKey(_aapl))
+            {
+                throw new Exception("AAPL TradeBar data added to algorithm before constituent universe selection took place");
+            }
+
+            if (data.Bars.Count == 1 && data.Bars.ContainsKey(_spy))
+            {
+                return;
+            }
+            
+            if (data.Bars.Count != 0 && !data.Bars.ContainsKey(_aapl))
+            {
+                throw new Exception($"Expected AAPL TradeBar data in OnData on {UtcTime:yyyy-MM-dd HH:mm:ss}");
+            }
+
+            if (!Portfolio.Invested)
+            {
+                SetHoldings(_aapl, 0.5m);
```

Comment:
Think would be nice to add an example algorithm which follows the weight of the ETF by some factor